### PR TITLE
Refactor STH metric updates

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This change introduces a new `Instance` type that incorporates information on a running log/mirror instance. It also factors out the code that keeps updating STH metrics into an `Instance` method, which can be reused by other CTFE implementations. Finally, this loop can now also update STHs for mirror logs.

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
PTAL. Might be convenient to do it commit-by-commit.

File: trillian/ctfe/sth.go [hunk 1]
```diff
@@ -35,6 +35,7 @@ var remoteQuotaCtxKey = contextKey("quotaUser")
 type MirrorSTHStorage interface {
 	// GetMirrorSTH returns an STH of TreeSize <= maxTreeSize. It does best
 	// effort to maximize the returned STH's TreeSize and/or Timestamp.
+	// If maxTreeSize < 0 then returns the latest known STH.
```
Comment (pavelkalinnikov) [inline trillian/ctfe/sth.go]: Oups, this line is not intended. I will remove it.

**Comment by Martin2112** [comment]
Ah great. I was in the middle of refactoring this to get rid of the deprecated signed log root stuff.

**Comment by Martin2112** [comment]
Though I don't think there's much overlap - phew!

File: trillian/ctfe/handlers.go [hunk 2]
```diff
@@ -490,34 +503,16 @@ func addPreChain(ctx context.Context, li *logInfo, w http.ResponseWriter, r *htt
 	return addChainInternal(ctx, li, w, r, true)
 }
 
-// PingTreeHead retrieves a tree head for the given log, and updates the STH
```
Comment (Martin2112) [inline trillian/ctfe/handlers.go]: I was going to suggest moving this out of `handlers.go` as it's not really related.

Comment (pavelkalinnikov) [reply trillian/ctfe/handlers.go]: Yeah, I moved it to Trash. Does that look good?

Comment (Martin2112) [reply trillian/ctfe/handlers.go]: Couldn't be better.

File: trillian/ctfe/instance.go [hunk 3]
```diff
@@ -62,28 +63,43 @@ type InstanceOptions struct {
...
-// SetUpInstance sets up a log (or log mirror) instance using the provided
-// configuration, and returns a set of handlers for this log. This is the
-// standard way of setting up a log instance.
-func SetUpInstance(ctx context.Context, opts InstanceOptions) (*PathHandlers, error) {
-	handlers, _, err := SetUpInstanceAndGetter(ctx, opts)
-	return handlers, err
+// Instance is a set up log/mirror instance. It must be created with the
+// SetUpInstance call.
+type Instance struct {
+	Opts      InstanceOptions
```
Comment (pavelkalinnikov) [inline trillian/ctfe/instance.go]: This one can probably be removed for now, as it's a copy of `li.instanceOpts`, and so far not used externally.

Comment (Martin2112) [reply trillian/ctfe/instance.go]: OK I won't comment on the doc string then. TODO for removal in a followup?

File: trillian/ctfe/instance.go [hunk 4]
```diff
@@ -62,28 +63,43 @@ type InstanceOptions struct {
...
 }
 
-// SetUpInstanceAndGetter sets up a log (or log mirror) instance using the
-// provided configuration, and returns a set of handlers for this log and the
-// STHGetter being used for it.
-//
-// Note: This is for experiments with DNS serving and is not currently used
-// by CTFE. This may either go away or become an official thing in future.
-func SetUpInstanceAndGetter(ctx context.Context, opts InstanceOptions) (*PathHandlers, STHGetter, error) {
+// RunUpdateSTH regularly updates the internal STH for each log so our metrics
```
Comment (Martin2112) [inline trillian/ctfe/instance.go]: Maybe 'Instance' instead of 'internal'.

File: trillian/ctfe/instance.go [hunk 5]
```diff
@@ -62,28 +63,43 @@ type InstanceOptions struct {
...
+// stay up-to-date with any tree head changes that are not triggered by us.
+func (i *Instance) RunUpdateSTH(ctx context.Context, period time.Duration) {
+	c := i.Opts.Validated.Config
+	ticker := time.NewTicker(period)
+	glog.Infof("Start internal get-sth operations on %v (%d)", c.Prefix, c.LogId)
+	for t := range ticker.C {
+		glog.V(1).Infof("Tick at %v: force internal get-sth for %v (%d)", t, c.Prefix, c.LogId)
+		_, err := i.li.getSTH(ctx)
+		if err != nil {
```
Comment (Martin2112) [inline trillian/ctfe/instance.go]: Can combine with previous line.

Comment (pavelkalinnikov) [reply trillian/ctfe/instance.go]: Done.

File: trillian/ctfe/instance.go (hunk 4 above)
Comment (pavelkalinnikov) [reply trillian/ctfe/instance.go]: Done.

**Comment by Martin2112** [comment]
Still seems to be a build issue.

**Comment by pavelkalinnikov** [comment]
It's a coverage, not build issue.

**Comment by Martin2112** [comment]
Yeah I was looking at the previous travis run. Oh well.

**Pull request merged**